### PR TITLE
[Configuration]Support X509 client certificate in kube config file

### PR DESCRIPTION
**Pull request by ityuhui**
Load configuration for X.509 client certificate in kube config file (~/.kube/config)

/cc @brendandburns

### CODE REVIEW

File: kubernetes/config/kube_config.c [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+{
+    int rc = 0;
+    char *_basePath = strdup(basePath);
+    if (_basePath) {
+        *pBasePath = _basePath;
+    } else {
+        rc = -1;
+    }
+    return rc;
+
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:21]: delete newline

File: kubernetes/config/kube_config.c [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+#include <unistd.h>
+
+#define ENV_KUBECONFIGE "KUBECONFIG"
+#define ENV_HOME "HOME"
+#define KUBE_CONFIG_DEFAULT_LOCATION "%s/.kube/config"
+#define KUBE_CONFIG_TEMPFILE_NAME_TEMPLATE "/tmp/kubeconfig-XXXXXX"
+
+static int setBasePath(char **pBasePath, char *basePath)
+{
+    int rc = 0;
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:13]: is this necessary? modern c supports multi-return, so why not just:

```c
if (...) {
  return 0;
}
return -1;
```

File: kubernetes/config/kube_config.c [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    char tempfile_name_template[] = KUBE_CONFIG_TEMPFILE_NAME_TEMPLATE;
+    int fd = mkstemp(tempfile_name_template);
+    if (-1 == fd) {
+        fprintf(stderr, "%s: Creating temp file for kubeconfig failed with error [%s]\n", fname, strerror(errno));
+        return NULL;
+    }
+
+    if (-1 == write(fd, b64decode, decoded_bytes)) {
+        fprintf(stderr, "%s: Writing temp file failed with error [%s]\n", fname, strerror(errno));
+        close(fd);
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:44]: hrm, I kind of don't love the `close(fd)` in multiple places, it's vulnerable to later cut/paste bugs.

How about:

```c
int rc = write(...);
close(fd);
return (rc == -1 ? NULL : strdup(...));
```

File: kubernetes/config/kube_config.c [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    if (sslConfig->clientCertFile) {
+        kubeconfig_rm_tempfile(sslConfig->clientCertFile);
+    }
+    if (sslConfig->clientKeyFile) {
+        kubeconfig_rm_tempfile(sslConfig->clientKeyFile);
+    }
+    if (sslConfig->CACertFile) {
+        kubeconfig_rm_tempfile(sslConfig->CACertFile);
+    }
+
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:75]: delete empty line

File: kubernetes/config/kube_config.c [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    if (ca_file) {
+        free(ca_file);
+    }
+
+    return rc;
+}
+
+static int setApiKeys(list_t ** pApiKeys, kubeconfig_user_t * user)
+{
+    int rc = 0;
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:124]: as above wrt if this is needed?

```c
if (...) {
   ...
   return 0;
}
return -1;
```

File: kubernetes/config/kube_config.c [hunk 6]
```diff
@@ -0,0 +1,309 @@
+#include "kube_config.h"
+#include "kube_config_yaml.h"
+#include <errno.h>
+#include <unistd.h>
+
+#define ENV_KUBECONFIGE "KUBECONFIG"
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:6]: `ENV_KUBECONFIG` ? (why the trailing `E`)

File: kubernetes/config/kube_config.c [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+                    sprintf(configFileName, KUBE_CONFIG_DEFAULT_LOCATION, homedir_env);
+                }
+            }
+        }
+    }
+
+    return configFileName;
+}
+
+static kubeconfig_context_t *kubeconfig_get_current_context(kubeconfig_context_t ** contexts, int contexts_count, const char *current_context_str)
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:163]: `static const kubeconfig_context_t *` ?

File: kubernetes/config/kube_config.c [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+        if (0 == strcmp(contexts[i]->name, current_context_str)) {
+            current_context = contexts[i];
+            break;
+        }
+    }
+
+    return current_context;
+}
+
+static kubeconfig_cluster_t *kubeconfig_get_current_cluster(kubeconfig_cluster_t ** clusters, int clusters_count, const char *cluster_name)
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:182]: return a const pointer?

File: kubernetes/config/kube_config.c [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+{
+    kubeconfig_cluster_t *current_cluster = NULL;
+
+    if (NULL == clusters || NULL == cluster_name) {
+        return NULL;
+    }
+
+    int i = 0;
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:190]: nit: 

```
for(int i = 0; i < clusters_count; i++) {
...
}
```

(here and everywhere)

File: kubernetes/config/kube_config.c [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+static kubeconfig_user_t *kubeconfig_get_current_user(kubeconfig_user_t ** users, int users_count, const char *user_name)
+{
+    kubeconfig_user_t *current_user = NULL;
+
+    if (NULL == users || NULL == user_name) {
+        return NULL;
+    }
+
+    int i = 0;
+    for (i = 0; i < users_count; i++) {
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:210]: optional suggestion:

I feel like if you made the objects union types you might be able to eliminate all of this redundant (but slightly different) code for each of these types.

File: kubernetes/config/kube_config.c [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+{
+    static char fname[] = "load_kube_config()";
+
+    int rc = 0;
+
+    kubeconfig_context_t *current_context = NULL;
+    kubeconfig_cluster_t *current_cluster = NULL;
+    kubeconfig_user_t *current_user = NULL;
+
+    kubeconfig_t *kubeconfig = kubeconfig_create();
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:230]: Does this need to be heap allocated? If you stack allocated it, you could remove all of the `goto end` statements, and just `return`

**Comment by brendandburns** [comment]
Looks great, I had a few comments.

Thanks!

File: kubernetes/config/kube_config.c (hunk 1 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:21]: OK.

File: kubernetes/config/kube_config.c (hunk 2 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:13]: OK.

File: kubernetes/config/kube_config.c (hunk 3 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:44]: OK.

File: kubernetes/config/kube_config.c (hunk 4 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:75]: OK.

File: kubernetes/config/kube_config.c (hunk 5 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:124]: OK.

File: kubernetes/config/kube_config.c (hunk 6 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:6]: Typo. I will fix it.

File: kubernetes/config/kube_config.c (hunk 7 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:163]: OK.

File: kubernetes/config/kube_config.c (hunk 8 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:182]: OK.

File: kubernetes/config/kube_config.c (hunk 9 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:190]: OK.

File: kubernetes/config/kube_config.c (hunk 10 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:210]: It's a very good idea. I ever want to use Macro to reduce the reduncant code. But obviously, union is better.

File: kubernetes/config/kube_config.c (hunk 11 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:230]: The structure ```kubeconfig_t``` (and its sub-fields such as kubeconfig_cluster_t, kubeconfig_context_t,etc...) incluces lots fields of  ```char *```, even though  kubeconfig_t (and its sub-fields such as kubeconfig_cluster_t, kubeconfig_context_t,etc...) is allocated in stack,  the ```char *``` fields are still allocated in heap. So we need call the function ```kubeconfig_free``` to free all memory before function returns.

Changing the  ```char *``` to ```char xxx[buffer_size]``` is also not a better solution.

File: kubernetes/config/kube_config.c (hunk 10 above)
Comment (ityuhui) [reply kubernetes/config/kube_config.c:210]: Hi @brendandburns 

Is the below design same with your thinking of union ?

```c
typedef struct kubeconfig_t {
    char *fileName;
    char *apiVersion;
    char *preferences;
    char *kind;
    char *current_context;
    kubeconfig_property_t **contexts;
    int  contexts_count;
    kubeconfig_property_t **clusters;
    int  clusters_count;
    kubeconfig_property_t **users;
    int  users_count;
} kubeconfig_t;

typedef struct kubeconfig_property_t {
    char *name;
    union {
        kubeconfig_cluster_t cluster;
        kubeconfig_user_t user;
        kubeconfig_context_t context;
    };
} kubeconfig_property_t;

typedef struct kubeconfig_cluster_t {
    char *server;
    char *certificate_authority_data;
} kubeconfig_cluster_t ;

typedef struct kubeconfig_context_t {
    char *cluster;
    char *user;
} kubeconfig_context_t ;

typedef struct kubeconfig_user_t {
    char *client_certificate_data;
    char *client_key_data;
    int insecure_skip_tls_verify;
...
} kubeconfig_user_t;
```

**Comment by ityuhui** [comment]
Hi @brendandburns 

The code is updated.  Could you please review the new code change ?

File: kubernetes/config/kube_config.c (hunk 10 above)
Comment (brendandburns) [reply kubernetes/config/kube_config.c]: Yeah, that looks about like what I was thinking about.

**Comment by ityuhui** [comment]
Thank you @brendandburns

**Pull request merged**